### PR TITLE
Tweaks for opening datasets

### PR DESCRIPTION
**Pull request by tsupinie**
I tweaked the open_dataset() and open_mfdataset() functions for better performance with the PyNIO engine. 
1. I use a lot of HDF4 files with what I guess I'll call "malformed names," where the file name does not end with ".hdf". PyNIO is able to figure out the format of the file, but it prints an annoying warning. As far as I can tell, the only way to shut off the warning is to tell it the format of the file, which I've added a `format` option for.
2. I added an option called `only_variables` which specifies which variables to load from the dataset in the event you don't want to load all variables. Say, for example, I have a dataset with 47 variables in it, but I only need 3 of them. If the data are not cached, then only loading the 3 I need cuts the I/O time in half. If they _are_ cached, then loading only the 3 takes 20% of the time to load the full dataset. The `only_variables` option behaves pretty similarly to `drop_variables`. The default is to load all variables.

### CODE REVIEW

**Comment by mathause** [comment]
I like (2) - I once needed this functionality and had to infer the `drop_vars`.

````
def get_drop_vars(variables, ds):

    if isinstance(variables, basestring):
        variables = [variables]
    else:
        variables = list(variables)

    # all variables
    drop_var = set(ds.variables.keys())

    # do not drop: coordinates, variables and time bounds
    keep_var = ds.coords.keys() + variables + ['time_bounds', 'time_bnds']

    drop_var.difference_update(keep_var)

    return drop_var`

**Comment by fmaussion** [comment]
Maybe `keep_variables` would be a more appropriate name? ('keep' vs 'drop').

File: xarray/backends/api.py [hunk 1]
```diff
@@ -81,7 +81,8 @@ def check_name(name):
 def open_dataset(filename_or_obj, group=None, decode_cf=True,
                  mask_and_scale=True, decode_times=True,
                  concat_characters=True, decode_coords=True, engine=None,
-                 chunks=None, lock=None, drop_variables=None):
+                 chunks=None, lock=None, drop_variables=None, 
+                 only_variables=None, format=''):
```
Comment (shoyer) [inline xarray/backends/api.py]: The default should be `format=None`.

File: xarray/backends/api.py [hunk 2]
```diff
@@ -135,6 +136,14 @@ def open_dataset(filename_or_obj, group=None, decode_cf=True,
         A variable or list of variables to exclude from being parsed from the
         dataset. This may be useful to drop variables with problems or
         inconsistent values.
+    only_variables: string or iterable, optional
+        A variable or list of variables to load from the dataset. This is
+        useful if you don't need all the variables in the file and don't want
+        to spend time loading them. Default is to load all variables.
+    format: string, optional
+        The format of the file to open (PyNIO engine only). This may be useful
```
Comment (shoyer) [inline xarray/backends/api.py]: I would simply say "currently only used by the PyNIO engine". We might eventually use this later, e.g., for selecting the fastest engine to use when open netCDF3 files (https://github.com/pydata/xarray/issues/887).

**Comment by shoyer** [comment]
This looks like nice functionality!

The main thing it needs is tests. Let me know if you have any questions about how to go about writing those.

I also like the name `keep_variables` better than than `only_variables`.

**Comment by jhamman** [comment]
@tsupinie - are you interested in finishing this up? Beyond a few changes requested by @shoyer, we just need some test coverage and a whatsnew entry.

**Comment by max-sixty** [comment]
Would anyone like to take this up? I was going to close as stale, but it's fairly close!

**Comment by mathause** [comment]
Once #2732 is merged (#2380) the first issue from the list is resolved. Id' still be interested to get No 2 in.Maybe it would be better to open a new PR?

**Comment by mathause** [comment]
I am closing this as stale. Half of this OR is implemented (#2732) the other half is  #1754.

**Pull request closed without merging**